tools/shared.py

```diff
@@ -772,6 +772,54 @@ def clean_temp():
     prepare_to_clean_temp(EMSCRIPTEN_TEMP_DIR) # this global var might change later
   return EMSCRIPTEN_TEMP_DIR
 
+class Enum(set):
+  def __getattr__(self, key):
+    if key in self:
+      return key
+
+WarningType = Enum(['ABSOLUTE_PATHS', 'SEPARATE_ASM', 'ALMOST_ASM'])
```

Comment:
I liked the idea of predefining warning types if only to make it easier to find documentation within the code on what warnings exist. De-optimizing to ASM_JS=2 (ALMOST_ASM) appears in two different places, but it might be that each would prefer to add their own customized message, e.g. "ALLOW_MEMORY_GROWTH caused this warning to display because its not supported in the asm.js spec". The WarningManager could supply a default message and warn will use the default message if parameter message=None. 

For option 1, I would need to add another API on WarningManager to disable_warning (and also enabled_warning for completeness) to handle some warnings being disabled by default (absolute paths). Default warning status will need to be initialized on the WarningManager prior to emcc calling capture_warnings. It seems cleaner for the WarningManager to manage the default status of all warnings vs having a block at the top of emcc before cmd arg parsing.